File: app/src/main/java/com/lusle/android/soon/View/Main/Setting/PreferenceFragment.java
package com.lusle.android.soon.View.Main.Setting;

import android.content.Context;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.provider.SearchRecentSuggestions;
import android.widget.Toast;

import com.google.gson.Gson;
import com.lusle.android.soon.MySuggestionProvider;
import com.lusle.android.soon.R;
import com.mukesh.countrypicker.Country;
import com.mukesh.countrypicker.CountryPicker;
import com.mukesh.countrypicker.listeners.OnCountryPickerListener;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

public class PreferenceFragment extends PreferenceFragmentCompat {

    private SharedPreferences preferences;

    public static PreferenceFragment newInstance() {
        PreferenceFragment fragment = new PreferenceFragment();
        return fragment;
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);

        preferences = PreferenceManager.getDefaultSharedPreferences(context);
    }

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        addPreferencesFromResource(R.xml.preferences);

        setRegionPreference();
        setClearSearchHistory();
    }

    private void setRegionPreference() {
        Preference regionPref = findPreference(getString(R.string.key_region));
        Country currC = new Gson().fromJson(preferences.getString(getString(R.string.key_region), ""), Country.class);
        if(currC == null) currC = new CountryPicker.Builder().with(getContext()).build().getCountryFromSIM();
        regionPref.setSummary(currC.getName());
        regionPref.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                CountryPicker.Builder builder =
                        new CountryPicker.Builder().with(getContext())
                                .listener(new OnCountryPickerListener() {
                                    @Override
                                    public void onSelectCountry(Country country) {
                                        preference.setSummary(country.getName());
                                        SharedPreferences.Editor editor = preferences.edit();
                                        editor.putString(getString(R.string.key_region), new Gson().toJson(country));
                                        editor.apply();
                                    }
                                });

                builder.build().showBottomSheet((AppCompatActivity) getContext());
                return false;
            }
        });
    }

    private void setClearSearchHistory() {
        findPreference(getString(R.string.key_clear_search_history)).setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
                builder
                        .setTitle("알림")
                        .setMessage("삭제시 복수할 수 없습니다")
                        .setNegativeButton("취소", (dialog, which) -> dialog.dismiss())
                        .setPositiveButton("삭제", (dialog, which) -> {
                            SearchRecentSuggestions suggestions = new SearchRecentSuggestions(getContext(), MySuggestionProvider.AUTHORITY, MySuggestionProvider.MODE);
                            suggestions.clearHistory();
                            dialog.dismiss();
                            Toast.makeText(getContext(), "검색 기록이 삭제 되었습니다", Toast.LENGTH_SHORT).show();
                        });
                builder.show();
                return false;
            }
        });
    }
}

File: app/src/main/java/com/lusle/android/soon/View/Main/Setting/PreferenceFragment.kt
package com.lusle.android.soon.View.Main.Setting

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.provider.SearchRecentSuggestions
import androidx.appcompat.app.AlertDialog
import androidx.navigation.fragment.findNavController
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import com.google.android.gms.oss.licenses.OssLicensesMenuActivity
import com.google.gson.Gson
import com.lusle.android.soon.MySuggestionProvider
import com.lusle.android.soon.R
import com.lusle.android.soon.Util.CountryPickerDialog.CCPCountry
import com.lusle.android.soon.Util.CountryPickerDialog.CountryCodeDialog
import com.lusle.android.soon.Util.CountryPickerDialog.CountryCodePicker

/**
 * A simple [Fragment] subclass.
 */
public class PreferenceFragment : PreferenceFragmentCompat() {
    private lateinit var pref: SharedPreferences

    private val DEFAULT_NAME_CODE = "kr"
    override fun onAttach(context: Context) {
        super.onAttach(context)
        pref = PreferenceManager.getDefaultSharedPreferences(context)
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        addPreferencesFromResource(R.xml.preferences)
        setUpReleaseAlarmPreference()
        setUpCompanyAlarmPreference()
        setRegionPreference()
        setClearHistoryPreference()
        setUpAppInfoPreference()
        setLincensePreference()
    }

    private fun setLincensePreference() {
        findPreference<Preference>(getString(R.string.key_license))!!.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            startActivity(Intent(requireContext(), OssLicensesMenuActivity::class.java))
            false
        }
    }

    private fun setUpAppInfoPreference() {
        findPreference<Preference>(getString(R.string.key_app_info))!!.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            findNavController().navigate(R.id.action_preferenceFragment_to_appInfoFragment)
            false
        }
    }

    private fun setUpCompanyAlarmPreference() {
        findPreference<Preference>(getString(R.string.key_company_alarm))!!.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            findNavController().navigate(R.id.action_preferenceFragment_to_companyAlarmSettingFragment)
            false
        }
    }

    private fun setUpReleaseAlarmPreference() {
        findPreference<Preference>(getString(R.string.key_release_alarm))!!.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            findNavController().navigate(R.id.action_preferenceFragment_to_releaseAlarmSettingFragment)
            false
        }
    }

    private fun setUpRegion() {
        val country = CCPCountry.getCountryForNameCodeFromLibraryMasterList(context, CountryCodePicker.Language.KOREAN, DEFAULT_NAME_CODE)
        pref.edit().putString(requireContext().getString(R.string.key_region), Gson().toJson(country)).apply()
    }

    private fun setClearHistoryPreference() {
        findPreference<Preference>(getString(R.string.key_clear_search_history))!!.onPreferenceClickListener = Preference.OnPreferenceClickListener {
            val builder = AlertDialog.Builder(requireContext())
            builder.setTitle("알림").setMessage("삭제시 복구할 수 없습니다").setPositiveButton("삭제") { dialog, _ ->
                val suggestions = SearchRecentSuggestions(context,
                        MySuggestionProvider.AUTHORITY, MySuggestionProvider.MODE)
                suggestions.clearHistory()
                dialog.dismiss()
            }.setNegativeButton("취소") { dialog, _ -> dialog.dismiss() }
            builder.show()
            false
        }
    }

    private fun setRegionPreference() {
        findPreference<Preference>(getString(R.string.key_region))!!.onPreferenceClickListener = Preference.OnPreferenceClickListener { preference ->
            val countryCodePicker = CountryCodePicker(context)
            val country = Gson().fromJson(
                    pref.getString(getString(R.string.key_region), null),
                    CCPCountry::class.java
            )
            countryCodePicker.setDialogKeyboardAutoPopup(false)
            countryCodePicker.selectedCountry = country
            countryCodePicker.changeDefaultLanguage(CountryCodePicker.Language.KOREAN)
            countryCodePicker.setOnCountryChangeListener {
                val country = countryCodePicker.selectedCountry
                preference.summary = country.name
                pref.edit().putString(getString(R.string.key_region), Gson().toJson(country)).apply()
            }
            CountryCodeDialog.openCountryCodeDialog(countryCodePicker, pref.getString(getString(R.string.key_region), null))
            false
        }
        if (pref.getString(getString(R.string.key_region), null) == null) setUpRegion()
        val countryJson = pref.getString(getString(R.string.key_region), "")
        findPreference<Preference>(getString(R.string.key_region))!!.summary = Gson().fromJson(countryJson, CCPCountry::class.java).name
    }
}